File: dailyQuestion/2020-07/07-07/python/solution_bfs.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def hasPathSum(root: TreeNode, sum: int) -> bool:
    if not root :
        return False
    queue = [root]
    while queue :
        root = queue.pop()
        if not root.left and not root.right :
            if root.val == sum :
                return True
        else :
            if root.left :
                item = root.left
                item.val = item.val + root.val
                queue.append(item)
            if root.right :
                item = root.right
                item.val = item.val + root.val
                queue.append(item)
    return False

if __name__ == "__main__" :
    node = TreeNode(5)
    node.left = TreeNode(4)
    node.left.left = TreeNode(11)
    node.left.left.left = TreeNode(7)
    node.left.left.right = TreeNode(2)
    node.right = TreeNode(8)
    node.right.right = TreeNode(4)
    node.right.right.right = TreeNode(1)
    result = hasPathSum(node,22)
    print(result)

File: dailyQuestion/2020/2020-07/07-07/golang/solution_bfs.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hasPathSum(root *TreeNode, sum int) bool {
	if root == nil {
		return false
	}
	queue := []*TreeNode{root}
	for len(queue) != 0 {
		root = queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		if root.Left == nil && root.Right == nil {
			if root.Val == sum {
				return true
			}
		} else {
			if root.Left != nil {
				item := root.Left
				item.Val = item.Val + root.Val
				queue = append(queue, item)
			}
			if root.Right != nil {
				item := root.Right
				item.Val += root.Val
				queue = append(queue, item)
			}
		}
	}

	return false
}

func main() {
	node := &TreeNode{Val: 4}
	result := hasPathSum(node,22)
	fmt.Println(result)
}
